aws/resource_aws_ec2_capacity_reservation_test.go

```diff
@@ -0,0 +1,222 @@
+		Steps: []resource.TestStep{
+			{
+				Config: testAccEc2CapacityReservationConfig,
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAWSEc2CapacityReservation_basic(t *testing.T) {
+	var cr ec2.CapacityReservation
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckEc2CapacityReservationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccEc2CapacityReservationConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckEc2CapacityReservationExists("aws_ec2_capacity_reservation.default", &cr),
+					resource.TestCheckResourceAttr("aws_ec2_capacity_reservation.default", "instance_type", "t2.micro"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSEc2CapacityReservation_endDate(t *testing.T) {
+	var cr ec2.CapacityReservation
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckEc2CapacityReservationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccEc2CapacityReservationConfig_endDate,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckEc2CapacityReservationExists("aws_ec2_capacity_reservation.default", &cr),
+					resource.TestCheckResourceAttr("aws_ec2_capacity_reservation.default", "end_date", "2019-10-31T07:39:57Z"),
+					resource.TestCheckResourceAttr("aws_ec2_capacity_reservation.default", "end_date_type", "limited"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSEc2CapacityReservation_tags(t *testing.T) {
+	var cr ec2.CapacityReservation
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckEc2CapacityReservationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccEc2CapacityReservationConfig_tags,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckEc2CapacityReservationExists("aws_ec2_capacity_reservation.default", &cr),
+					resource.TestCheckResourceAttr("aws_ec2_capacity_reservation.default", "instance_type", "t2.micro"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckEc2CapacityReservationExists(resourceName string, cr *ec2.CapacityReservation) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Not found: %s", resourceName)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No ID is set")
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).ec2conn
+		resp, err := conn.DescribeCapacityReservations(&ec2.DescribeCapacityReservationsInput{
+			CapacityReservationIds: []*string{aws.String(rs.Primary.ID)},
+		})
+
+		if err != nil {
+			return fmt.Errorf("DescribeCapacityReservations error: %v", err)
+		}
+
+		if len(resp.CapacityReservations) > 0 {
```

Comment:
This function should verify that the reservation is in an active or pending state. 👍 I'd suggest inverting this conditional logic at the same time:

```go
if len(resp.CapacityReservations) == 0 {
  return fmt.Errorf("EC2 Capacity Reservation (%s) not found", rs.Primary.ID)
}

reservation := resp.CapacityReservations[0]

if aws.StringValue(reservation.State) != ec2.CapacityReservationStateActive && aws.StringValue(reservation.State) != ec2.CapacityReservationStatePending {
  return fmt.Errorf("EC2 Capacity Reservation (%s) found in unexpected state: %s", rs.Primary.ID, aws.StringValue(reservation.State))
}

*cr = *reservation
return nil
```